File: step05listview/src/main/java/com/example/step05listview/MainActivity.java
package com.example.step05listview;

import android.app.AlertDialog;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity
                                implements AdapterView.OnItemClickListener,
                                            AdapterView.OnItemLongClickListener {
    List<String> names;
    ArrayAdapter<String> adapter;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //sample 데이터
        names=new ArrayList<>();
        names.add("김구라");
        names.add("해골");
        names.add("원숭이");
//        for (int i=0; i<100; i++){
//            names.add("아무개 "+(i+1));
//        }

        //ListView에 연결할 아답터
        adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1, names);
        //ListView의 참조값 얻어오기
        ListView listView = findViewById(R.id.listView);
        //ListView에 아답터 연결하기
        listView.setAdapter(adapter);

        // Button에 리스너 등록하기
        Button addBtn = findViewById(R.id.addBtn);
        addBtn.setOnClickListener(v->{
            // inputName의 참조값
            EditText inputName = findViewById(R.id.inputName);
            // inputName의 문자열
            String name = inputName.getText().toString();
            // names 에 추가하고
            names.add(name);
            // 아답터에 names 가 변경되었다고 알린다.
            adapter.notifyDataSetChanged();
            // EditText 에 입력한 내용 지우기
            inputName.setText("");
            // 마지막 추가한 위치까지 부드럽게 스크롤하기
            int position = adapter.getCount();   //전체 아이템의 갯수
            listView.smoothScrollToPosition(position);
        });
        //ListView 에 AdapterView.ItemClickListener 등록하기
        listView.setOnItemClickListener(this);
        listView.setOnItemLongClickListener(this);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        /*
            parent => ListView
            view => cell view (TextView)
            position => 클릭한 cell의 인덱스
            id => 클릭한 cell의 item이 id가 있다면, 해당 값이 전달된다. (없으면 index가 전달됨)
         */
        Log.d("MainActivity","position : "+position);
        // 클릭한 cell에 출력된 이름
        String name=names.get(position);
        // Toast메세지로 출력해보기
        Toast.makeText(this,name,Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage("삭제할까요?")
                .setPositiveButton("네", (dialog, which) -> {
                    //Long Click 된 cell 의 index => position
                    names.remove(position);
                    //UI 업데이트
                    adapter.notifyDataSetChanged();
                })
                .setNegativeButton("아니요", null)
                .create()
                .show();
        return false;
    }
}

File: step08galleryexample/src/main/java/com/example/step08galleryexample/MainActivity.kt
package com.example.step08galleryexample

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.step08galleryexample.databinding.ActivityMainBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

class MainActivity : AppCompatActivity() , OnItemClickListener {
    lateinit var binding:ActivityMainBinding
    lateinit var list:MutableList<GalleryDto>
    lateinit var adapter:GalleryAdapter
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding=ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        //모델
        list = mutableListOf<GalleryDto>()
        //ListView 에 연결할 Custom Adapter [라이브러리를 가져다 사용 할 예정]
        adapter= GalleryAdapter(this, R.layout.listview_cell, list)
        //ListView 에 Adapter 연결하기
        binding.listView.adapter=adapter    //binding.listView.setAdapter(adapter)
        //gallery 데이터를 서버로 부터 받아온다.
        lifecycleScope.launch (Dispatchers.Main){
            val url="http://192.168.0.68:9000/boot09/api/galleries"
            // [{"num":x,"writer":"xxx", ...}]
            val json=makeHttpRequest(url)
            // 응답받은 json 문자열을 이용해서 ListView 에 Gallery 출력하기
            printToListView(json)
        }
        //ListView 에 아이템 클릭 리스너 등록
        binding.listView.onItemClickListener = this
    }



    // json 문자열을 파싱해서 ListView 에 정보 출력하기
    fun printToListView (json:String){
        val arr=JSONArray(json)
        //반복문 돌면서
        for(i in 0 until arr.length()){
            // i번째 JSONObject 객체를 얻어와서
            val tmp=arr.getJSONObject(i)
            //GalleryDto 객체에 정보를 담고
            val dto = GalleryDto()
            dto.num = tmp.getInt("num")
            dto.writer = tmp.getString("writer")
            dto.caption = tmp.getString("caption")
            dto.regdate = tmp.getString("regdate")
            dto.imagePath = "http://192.168.0.68:9000/boot09/upload/images/"+
                    tmp.getString("saveFileName")
            //모델에 추가하고
            list.add(dto)
        }
        //모델이 변경되었다고 adapter에 알린다.
        adapter.notifyDataSetChanged()
    }

    suspend fun makeHttpRequest(url:String):String {
        /*
        *   IO(InputOutput) 입출력 처리를 하기에 적합한 스레드를 시작시켜서 작업을 한다.
        *   withContext(){} 블럭 내부는 UI 스레드가 아니다.
        *   블럭의 가장 아랫부분에 남긴 값이 리턴된다.
        */
        val result = withContext(Dispatchers.IO){
            // 여기도 새로운 스레드
            //문자열을 누적할 객체
            //StringBuilder builder = new StringBuilder();
            val builder = StringBuilder();

            try{
                //요청 url을 생성자의 인자로 전달하면서 URL 객체를 생성한다.
                val url = URL(url)
                //URLConnection 객체를 원래 type(자식 type) 으로 casting 해서 받는다.
                val conn = url.openConnection() as HttpURLConnection;
                if(conn != null){
                    conn.setConnectTimeout(20000);  // 응답을 기다리는 최대 대기 시간
                    conn.setRequestMethod("GET");   // 요청 메소드 설정 (Default 는 GET
                    //conn.setUseCaches(false);       // 캐시 사용 여부
                    conn.useCaches=false
                    //응답 코드를 읽어온다.
                    //int responseCode = conn.getResponseCode();
                    val responseCode = conn.responseCode
                    if(responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면(200)
                        //문자열을 읽어들일 수 있는 객체의 참조값 얻어오기
                        val br = BufferedReader(InputStreamReader(conn.inputStream));
                        //반복문 돌면서
                        while(true){
                            //문자열을 한줄 씩 읽어 들인다.
                            val line = br.readLine();
                            if(line==null)break;
                            // StringBuilder 객체에 누적시키기
                            builder.append(line);
                        }
                    }
                }

            }catch (e:Exception){
                e.message?.let{ Log.e("MainActivity", it)}
            }//try~catch

            //누적된 문자열이 리턴된다.
            builder.toString()// 이것이 withContext...return 값
        }
        return result
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        //ListView 의 아이템을 클릭하면 여기가 호출된다.
        //Java 에서 class type 은 클래스명 . class
        //kotlin 에서 class type 은 클래스명 :: class.java
        intent = Intent(this, DetailActivity::class.java)
        //intent 객체에 DetailActivity 에 전달할 정보를 담는다.
        intent.putExtra("id", id)   // id 라는 키값으로 Long type 을 담는다.
        //액티비티 이동
        startActivity(intent)
    }
}